xds/internal/server/listener_wrapper.go

```diff
@@ -113,20 +142,75 @@ type listenerWrapper struct {
 	// using an rw lock here is that this field will be read by all connections
 	// during their server-side handshake (in the hot path), but writes to this
 	// happen rarely (when we get a Listener resource update).
-	mu                 sync.RWMutex
-	filterChains       []*xdsclient.FilterChain
-	defaultFilterChain *xdsclient.FilterChain
+	mu           sync.RWMutex
+	filterChains *xdsclient.FilterChainManager
 }
 
 // Accept blocks on an Accept() on the underlying listener, and wraps the
 // returned net.connWrapper with the configured certificate providers.
 func (l *listenerWrapper) Accept() (net.Conn, error) {
-	c, err := l.Listener.Accept()
-	if err != nil {
-		return nil, err
+	var retries int
+	for {
+		conn, err := l.Listener.Accept()
+		if err != nil {
+			// Temporary() method is implemented by certain error types returned
+			// from the net package, and it is useful for us to not shutdown the
+			// server in these conditions. The listen queue being full is one
+			// such case.
+			if ne, ok := err.(interface{ Temporary() bool }); ok && ne.Temporary() {
```

Comment:
Invert to return on err and then outdent the more complex code?

Reply:
Done.